---
title: "Weekly update - 19.02.2021"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors=FALSE)
library(knitr)
library(GenomicRanges)
library(rtracklayer)
library(data.table)
library(kableExtra)
library(magrittr)

```

```{r results="asis", include=FALSE}
cat("
<style>
caption {
      color: red;
      font-weight: bold;
      font-size: 1.0em;
    }
</style>
")
```

## Subset GWAS summary statistics based on SNPs in the baseline. Re-run partitioned heritability analysis. #plot results, in comparison w/previous results

Based on the genomic regions in the baseline model, 123,438 SNPs are filtered out from each summary statistics file. Only the SNPs present in the baseline model are kept. Then, partitioned heritability analysis is repeated. Heritability enrichment and heritability proportion results from initial and second run are compared below:

```{r filter sumstats, echo=FALSE}

baseline_bed = read.table("/data/clusterfs/lag/users/gokala/enigma-evol/partherit/beds/allbed/all.bed")
baseline_bed_gr = makeGRangesFromDataFrame(baseline_bed, seqnames.field = "V1", start.field = "V2", end.field = "V3")

sumstats = list.files("/data/clusterfs/lag/users/gokala/enigma-evol/ancreg/replication_v1", pattern = "surfaceDK", full.names = T)

outDir = "/data/clusterfs/lag/users/gokala/enigma-evol/ancreg/baseline_subset/replication_v1/"
i=sumstats[12]
for (i in sumstats) {

    load(i) # load Rdata summary statistics
    subset = subsetByOverlaps(mergedGR, baseline_bed_gr) # subset summary statistics based on baseline bed file
    tmp_sumstat = mcols(subset) # convert Granges object to a data frame
    colnames(tmp_sumstat)[c(2,3,10)] = c("A1", "A2", "CHR") # correct column names A1.x, A2.x, CHR.y
    write.table(tmp_sumstat,paste0(outDir,sub(pattern = "(.*)\\..*$", replacement = "\\1", basename(i)),".txt"), quote = F)
    
    print(i)
    print(paste0(length(mergedGR)-length(subset)," SNPs are filtered out based on the baseline"))

}

```

## Overlap between annotation pairs

```{r annotation overlap, echo=FALSE}
#"foverlap" function is taken from Else Eising's script. Finds overlap between intervals, in this case genomic annotations.

fbeds = "/data/clusterfs/lag/users/gokala/enigma-evol/partherit/beds/"
beds = list.files("/data/clusterfs/lag/users/gokala/enigma-evol/partherit/beds/",pattern = "\\.")

summary_df = as.data.frame(matrix(0,ncol=length(beds),nrow=length(beds)))
rownames(summary_df) = beds
colnames(summary_df) = beds

for (i in beds) {
  tmp_bed = read.table(paste0(fbeds,i))
  tmp_bed = tmp_bed[,c(1,2,3)]
  colnames(tmp_bed) = c("chr", "start", "end")
  tmp_bed$length = (tmp_bed$end - tmp_bed$start)
  setDT(tmp_bed)
  setkey(tmp_bed, chr, start, end)

    for (j in beds) {
    tmp_bed_2 = read.table(paste0(fbeds,j))
    tmp_bed_2 = tmp_bed_2[,c(1,2,3)]
    colnames(tmp_bed_2) = c("chr", "start", "end")
    tmp_bed_2$length = (tmp_bed_2$end - tmp_bed_2$start)
    setDT(tmp_bed_2)
    setkey(tmp_bed_2, chr, start, end)
        
    overlap = foverlaps(tmp_bed, tmp_bed_2, by.x=c("chr", "start", "end"), type="any", mult="all", nomatch = 0)
    overlap$max_start = pmax(overlap$start,overlap$i.start)
    overlap$min_end = pmin(overlap$end,overlap$i.end)
    overlap$overlap_length = overlap$min_end - overlap$max_start
    summary_df[i,j] = sum(overlap$overlap_length)

    }
}

annotations = c("Human Gained Enhancers - 12 PCW (Frontal)",
                "Human Gained Enhancers - 12 PCW (Occipital)",
                "Human Gained Enhancers - 7 PCW",
                "Human Gained Enhancers - 8.5 PCW",
                "Human Gained Enhancers - Adult human vs. chimpanzee (prefrontal cortex)",
                "Human Gained Promoters - Adult human vs. chimpanzee (prefrontal cortex)",
                "Human Accelerated Regions",
                "Human Gained Enhancers - Adult human vs. macaque (prefrontal cortex)",
                "Human Gained Promoters - Adult human vs. macaque (prefrontal cortex)",
                "Neanderthal Lineage Depleted Regions",
                "Neanderthal Introgressed SNPs",
                "Selective Sweeps")

rownames(summary_df) = annotations
colnames(summary_df) = annotations

# colorize the diagonal elements in table with non-ints
for (i in 1:(ncol(summary_df))) {
  summary_df[i,i] = cell_spec(summary_df[i,i], bold=T, background = "lightgray")
}

overlap_table = kable(summary_df, caption = "Overlapping sequence length for each annotation pair (bp)", align = "c", escape = F, table.attr = "style = \"color: black;\"") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, html_font = "Cambria", font_size = 12)
overlap_table

#save_kable(overlap_table,"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/partitioned_heritability/plots/annotation_overlap_table.png")
  
```

## Re-run partitioned heritability for 23andMe. #plot previous and re-run of heritability results


```{r pressure, echo=FALSE}


```

sessionInfo()