#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)
#
# This script will...
# 1- Merge clumped files.
# 2- Read control variants generated by SNPsnap.
# 3- Run plink on control variants to expand control sets.
# 4- Pick random variants in LD with controls.
#
# Run this script with R 4.0.3, so that biomaRt works
# properly.
#
# Gokberk Alagoz
# created on: 02.06.2021

#####

library(biomaRt)
library(data.table)
options(stringsAsFactors=FALSE)

#--------------------------------------------
# PATHS

clumpedDir="/data/clusterfs/lag/users/gokala/enigma-evol/selection_analysis/clumped_sumstats/european_lr/clumped_sumstats_p10e-5/right"
#"/data/clusterfs/lag/users/gokala/enigma-evol/selection_analysis/clumped_sumstats/european_lr/clumped_sumstats_p10e-5/left"
genotypeFile = "/data/workspaces/lag/shared_spaces/Resource_DB/1KG_phase3/GRCh37/plink/1KG_phase3_GRCh37_EUR_nonFIN_allchr"
controlVars = "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/selection_analysis/european_lr/SNPsnap_eur_right/matched_snps_annotated_subset.filtered.txt"
outDir= "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/selection_analysis/european_lr/right"
leadSNPlist = "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/selection_analysis/european_hemave/working_snps.txt"
clumpedControls = "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/selection_analysis/european_hemave"

#--------------------------------------------
# FUNCTIONS

#assure_dir_exists #TODO

#check_for_plink_error #TODO

make_CSA_region_list = function(clumpedDir) {
  # gather rsID and TOTAL info lead SNPs from
  # each summary stats
  clumpedSNPs = data.frame()
  for (i in dir(clumpedDir, full.names = T, pattern = "clumped")) {
      region = sapply(i, function (x) {unlist(strsplit(as.character(x),"_",fixed=TRUE))[9]})
      tmp_clump = read.table(i, header = TRUE)
      tmp_df = data.frame(SNP = tmp_clump$SNP, TOTAL = tmp_clump$TOTAL)
      clumpedSNPs = rbind(clumpedSNPs, tmp_df)
  }
  write.table(clumpedSNPs,paste0(clumpedDir,"/snpsnap_list/surface_area_all.txt"),row.names = F, col.names = F, quote = F)
}

make_CSA_region_list_only_rsIDs = function(clumpedDir) {
  # gather rsID info lead SNPs from
  # each summary stats, use for SNPsnap
  # control variant generation
  clumpedSNPs = data.frame()
  clumpedFiles = intersect(dir(clumpedDir,full.names = T,pattern = "surface"),dir(clumpedDir,full.names = T,pattern = "clumped"))
  for (i in clumpedFiles) {
    region = sapply(i, function (x) {unlist(strsplit(as.character(x),"_",fixed=TRUE))[9]})
    tmp_clump = read.table(i,header=TRUE)
    tmp_df = data.frame(SNP=tmp_clump$SNP)
    clumpedSNPs=rbind(clumpedSNPs,tmp_df)
  }
  write.table(clumpedSNPs,paste0(outDir,"/snpsnap_list/surface_area_right_rsIDs.txt"),row.names = F, col.names = F, quote = F)
}

#get_num_LD_buddies = function(clumpedDir, outDir) { #TODO this prints 2 extra lines with ":", fix it!
#  # get the rsIDs and LD buddy number from clumped file
#  clumpedFiles = intersect(dir(clumpedDir,full.names = T,pattern = "surface"),dir(clumpedDir,full.names = T,pattern = "clumped"))
#  for (i in clumpedFiles) {
#    file_name_full = sapply(i, function (x) {unlist(strsplit(as.character(x),".",fixed=TRUE))[1]})
#    file_name_base = sapply(file_name_full, function (x) {unlist(strsplit(as.character(x),"/",fixed=TRUE))[11]})
#    system(paste0("awk '{print $1\":\"$4\"\t\"$3\"\t\"$6}' ", i, " > ", paste0(outDir,"/",file_name_base,"_LDbuddy_counts.txt"))) #TODO use column names instead of field numbers
#  }
#}

# Go run SNPsnap with the rsID lists!

## Run plink on 5k control variants to generate SNPs in LD with them
## do this ON GRID, it takes forever on lux13.
#run_plink_ld = function(genotypeFile, controlVars, outDir) {
#  # runs plink --ld in control snps
#  
#  controlVarsTable = read.table(controlVars, header = T)
#  #controlVarsTable = controlVarsTable[controlVarsTable$input_snp == args[1],] - what did I think here? Leaving this out until I remember why I added this.
#  
#  for (i in 1:nrow(controlVarsTable)) {
#    system(paste0("module load plink/1.9b6 \
#                   plink --bfile ", genotypeFile, " --r2 dprime --ld-snp ", 
#                  controlVarsTable$rsID[i], " --ld-window-kb 1000 --ld-window 99999 --ld-window-r2 0.9 --out ",
#                  outDir, "/", controlVarsTable$input_snp[i], "_", controlVarsTable$set[i], "_", 
#                  controlVarsTable$rsID[i]))
#    }
#}

extract_leadSNP_info = function(clumpedDir, leadSNPlist, controlVars, controlLDbuddies, outDir) {
  # exracts lead SNP info from .clumped files
  
  # Extract MARKER, rsID, LD-buddy number (TOTAL) and LD-buddies info.
  # from .clumped files.
  working_snps = read.table(leadSNPlist)
  surface = dir(clumpedDir, pattern = "surface", full.names = T)
  all_variants = data.frame(MARKER = character(),
                            rsID = character(),
                            TOTAL = character(),
                            LDbuds = character(),
                            stringsAsFactors = F)
  k = 1
  for (i in 1:length(grep(surface, pattern = "top10k.clumped", value = T))) {
    tmp_var = read.table(grep(surface, pattern = "top10k.clumped", value = T)[i], header = T)
    
    for (j in 1:nrow(tmp_var)){
      all_variants[k+(j-1),] = c(paste0(tmp_var$CHR[j],":", tmp_var$BP[j]),
                                 tmp_var$SNP[j], 
                                 tmp_var$TOTAL[j], 
                                 tmp_var$SP2[j])
    }
    k = k + j
  }
  
  # Filter all variants file, keep only variants
  # which passed SNPsnap filtering
  working_indeces =  data.frame(MARKER = character(),
                                rsID = character(),
                                TOTAL = character(),
                                LDbuds = character(),
                                stringsAsFactors = F)
  
  for (i in 1:nrow(working_snps)) {
    working_indeces[i,] = all_variants[which(all_variants$rsID == working_snps$V2[i]),]
  }
  
  # Remove (1)s from LD-buddy rsIDs
  
  working_indeces$LDbuds = gsub("\\(1\\)", "", working_indeces$LDbuds)
  
  # Create a df per lead SNP. Fill it with the following:
  # MARKER  rsID  TOTAL LDbuddy1  LDbuddy2  ... LDbuddyN
  # First line should have lead SNP info., and the rest of
  # the data frame will be filled with control SNPs and 
  # their LD-buddies.
  
  for (i in 1:nrow(working_indeces)) {
    
    controlLDbuddies = dir(paste0(clumpedControls,"/",working_indeces$MARKER[i]),
                           pattern = ".ld", full.names = T)
    
    # Create an empty matrix NxM
    # N = number of control variants from SNPsnap + 1 (leadSNP)
    # M = LD-buddy number of the lead SNP
    tmp_mat = matrix(nrow = length(controlLDbuddies)+1, ncol = as.numeric(working_indeces$TOTAL[i])+1)

    if (working_indeces$TOTAL[i]!=0) {
      # Feed in the lead SNP and its LD-buddies to
      # the first line.
      tmp_mat[1,1] = working_indeces[i,1]
      tmp_mat[1,2:ncol(tmp_mat)] = as.vector(strsplit(working_indeces[i,4], ","))[[1]]
      
      # Feed in the control variants and their LD-
      # buddies to the rest of the matrix.
      k = 2
      for (j in 1:(length(controlLDbuddies))) {
        tmp_controls = read.table(controlLDbuddies[j], header = T)
        tmp_mat[k,1] = paste0(tmp_controls$CHR_A[1],":",tmp_controls$BP_A[1])
        tmp_mat[k,2:ncol(tmp_mat)] = paste0(tmp_controls$CHR_B[1:working_indeces$TOTAL[i]],":",tmp_controls$BP_B[1:working_indeces$TOTAL[i]])
        k = k + 1
      }
      
      # Remove rows with NAs and SNPs with 2 markers (e.g. "12:123123;7:456456")
      
      tmp_df = as.data.frame(tmp_mat)
      tmp_df = na.omit(tmp_df)
      
      for (l in 1:nrow(tmp_df)) {
        if (any(grepl(";", tmp_df[l,]))) {
          tmp_df = tmp_df[-l,]
        }
      }
      
      # LD buddies of dear lead SNPs have only rsIDs. We have to convert them to 
      # markers to get the evol measures in the next step.
      
      for (m in 2:ncol(tmp_df)) {
        
        snpmart = useMart("ENSEMBL_MART_SNP", 
                          dataset = "hsapiens_snp", 
                          path="/biomart/martservice", 
                          host="https://grch37.ensembl.org")
        tmp_marker = getBM(attributes = c("chr_name","chrom_start","chrom_end"), 
                         filters = "snp_filter", 
                         values = tmp_df[1,m], #list(tmp_chr,tmp_pos,tmp_pos), 
                         mart = snpmart)
        tmp_df[1,m] = paste0(tmp_marker[1,1],":",tmp_marker[1,2])
      }
      
      # Remove rows with NAs
      #tmp_df = na.omit(tmp_df)
      
      write.table(tmp_df, file = paste0(outDir, "/", working_indeces[i,2], ".txt")
                  , quote = F, row.names = F, col.names = F, sep = "\t")
    }
    else {
      # If the lead SNP doesn't have any LD-buddies, things work
      # slightly differently.
      
      tmp_mat[1,1] = working_indeces[i,1]
      
      k = 2
      for (j in 1:(length(controlLDbuddies))) {
        tmp_controls = read.table(controlLDbuddies[j], header = T)
        tmp_mat[k,1] = paste0(tmp_controls$CHR_A[1],":",tmp_controls$BP_A[1])
        k = k + 1
      }
      
      # Remove rows with NAs and SNPs with 2 markers (e.g. "12:123123;7:456456")
      
      tmp_df = as.data.frame(tmp_mat)
      tmp_df = na.omit(tmp_df)
      
      for (l in 1:nrow(tmp_df)) {
        if (any(grepl(";", tmp_df[l,]))) {
          tmp_df = tmp_df[-l,]
        }
      }
            
      write.table(tmp_df, file = paste0(outDir, "/CSAregions/", working_indeces[i,2], ".txt")
                  , quote = F, row.names = F, col.names = F, sep = "\t")
    }
    
  }
  #TODO output files has lots of "NA:NA"s and corrupt elements. Find a way to fix it.
}

#--------------------------------------------
# MAIN

###
# Read in clumped SNPs from all summary stats
# and merge them.
###

#make_CSA_region_list_only_rsIDs(clumpedDir)

###
# Get number of SNPs in LD with each tag SNP
###

#get_num_LD_buddies(clumpedDir,outDir)

###
# Generate control variants with SNPsnap using
# guidelines in the paper.
###

#run_plink_ld(genotypeFile,controlVars,outDir)

#extract_leadSNP_info(clumpedDir, leadSNPlist, controlVars, controlLDbuddies, outDir)

#--------------------------------------------

sessionInfo()