---
title: "overlap_analysis_eQTL"
author: "Gokberk Alagoz"
date: "September 26, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
args = commandArgs(trailingOnly=TRUE)
options(stringsAsFactors=FALSE)
library(GenomicRanges)
library(biomaRt)
library(circlize)
library(tidyr)
library(knitr)
```

## R Markdown

```{r paths, echo=FALSE}
bedfile = args[1] # "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/resources/new_annotations/beds/GRCh37/neanDepRegions_hg19.sorted.bed"
sumstatsList = args[2] # "/data/clusterfs/lag/users/gokala/enigma-evol/data/european_lr/sumstats_txt_list.txt"
clumpedSumstatsDir = args[3] #"/data/clusterfs/lag/users/gokala/enigma-evol/eqtl/clumped_sumstats/european_lr"
outDir = args[4] # "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/eqtl/european_lr/results"
genotypeF = "/data/workspaces/lag/shared_spaces/Resource_DB/1KG_phase3/GRCh37/plink/1KG_phase3_GRCh37_EUR_nonFIN_allchr"
```

## Including Plots

```{r parse, echo=FALSE}

# eQTL data downloaded from PsychENCODE
feqtl = "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/old_results/eqtl/DER-08a_hg19_eQTL.significant.txt"

# Parse to get trait name
clumpfileloc = dir(clumpedSumstatsDir, pattern = ".clumped", full.names = T)

phenoname = paste0(sapply(clumpfileloc, function (x) {unlist(strsplit(x, "_", fixed = TRUE))[6]}), "_", 
                   sapply(clumpfileloc, function (x) {unlist(strsplit(x, "_", fixed = TRUE))[7]}), "_", 
                   sapply(clumpfileloc, function (x) {unlist(strsplit(x, "_", fixed = TRUE))[8]}))

annot_name = unlist(strsplit(unlist(strsplit(bedfile, "/", fixed=T))[15], ".", fixed=T))[1]

```

```{r fullSA_left}

# Full surface area (LEFT HEM.)
# Start by getting all the clumped SNPs only for full surface area

fullsurfind = which(phenoname=="surface_le_Full")
clump = read.table(clumpfileloc[fullsurfind],header=TRUE)

# Loop over all clumped SNPs
#i=1
for (i in 1:nrow(clump)) {
    
    # Find all SNPs in LD (r2>0.6) with each clumped SNP
  
    system(paste0("module load plink/1.9b6 \
                   plink --bfile ", genotypeF, " --r2 --ld-window-kb 10000 --ld-window 2000 --ld-window-r2 0.6 --ld-snp ",
                  clump$SNP[i], " --out tmpld_", annot_name))
    
    # Read in the LD calculated from plink
    
    LD = read.table(paste0("tmpld_",annot_name,".ld"),header=TRUE)
    
    # Turn into a genomic ranges object
    
    if (i==1) { 
       LDSNPs = GRanges(LD$CHR_B,IRanges(LD$BP_B,LD$BP_B),SNP = LD$SNP_B,indexSNP = LD$SNP_A)
    } else {
       LDSNPs = c(GRanges(LD$CHR_B,IRanges(LD$BP_B,LD$BP_B),SNP = LD$SNP_B,indexSNP = LD$SNP_A), LDSNPs)
    }
}
  
# Read in the eQTL data
eqtl = read.table(feqtl, header=TRUE)
eqtl.GR = GRanges(gsub("chr", "", eqtl$SNP_chr), IRanges(eqtl$SNP_start, eqtl$SNP_end))
mcols(eqtl.GR) = eqtl[, c(1:8, 12:15)]

# Read in the BED file containing the annotation file
annot = read.table(bedfile, header=FALSE)
annotGR = GRanges(gsub("chr", "", annot$V1), IRanges(annot$V2, annot$V3), type = annot$V4)
        
# Find overlaps
olap = findOverlaps(annotGR, LDSNPs)
regionalSA_vars = unique(LDSNPs$indexSNP[subjectHits(olap)])
regionalAnnot = LDSNPs[which(!is.na(match(LDSNPs$indexSNP, regionalSA_vars)))]

# Outputting the number of loci that overlap with an HGE
cat('Number of loci that overlap with ', annot_name, ' is: ', length(unique(regionalAnnot$indexSNP)), '\n')

olap2 = findOverlaps(eqtl.GR, regionalAnnot)
regionalSA_eqtl_vars  = unique(regionalAnnot$indexSNP[subjectHits(olap2)])
cat('Number of ', annot_name, ' overlapping loci that also have an eQTL: ', length(unique(regionalSA_eqtl_vars$indexSNP[subjectHits(olap2)])), '\n')
eqtlgenes = unique(eqtl.GR$gene_id[queryHits(olap2)])
        
# Remove the . annotation
eqtlgenes = sapply(eqtlgenes, function (x) {unlist(strsplit(x, ".", fixed=TRUE))[1]})

if (length(unique(regionalSA_eqtl_vars$indexSNP[subjectHits(olap2)])) > 0) {
  
  # Convert these genes to hgnc_id
  mart = useMart(biomart="ENSEMBL_MART_ENSEMBL",
                 dataset="hsapiens_gene_ensembl",
                 host="feb2014.archive.ensembl.org")
  geneannot = getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                    filters = "ensembl_gene_id",
                    values = eqtlgenes,
                    mart = mart)
  cat('These eQTLs impact ', length(which(geneannot == "protein_coding")), ' protein-coding eGenes\n')
  write.csv(geneannot, file = paste0(outDir, "/leftHem_SA_", annot_name, "_lr_eqtl_genes.csv"),
            row.names = FALSE, quote = FALSE)
  
} else {
  print("There are not any CSA-associated variant - eQTL overlap")
}

if (length(regionalSA_vars) > 0) {

  m = matrix(NA, nrow = length(regionalSA_vars), ncol = 3)
  d = as.data.frame(m)
  colnames(d) = c("olap_snps", "eqtl_snps", "region")
  
  d$olap_snps = regionalSA_vars
  d$eqtl_snps[match(regionalSA_eqtl_vars, d$olap_snps)] = "Yes"
  d$region = "full_leftHem"
  d$olap_snps[match(regionalSA_eqtl_vars, d$olap_snps)]
  write.csv(d, file = paste0(outDir, "/leftHem_SA_", annot_name, "_lr_olap_snps.csv"),
            row.names = FALSE, quote = FALSE)
} else {
  print("There are not any CSA-associated variants - annotation overlap")
}

```

```{r fullSA_right}

# Full surface area (RIGHT HEM.)
# Start by getting all the clumped SNPs only for full surface area

fullsurfind = which(phenoname == "surface_re_Full")
clump = read.table(clumpfileloc[fullsurfind], header = TRUE)

# Loop over all clumped SNPs

for (i in 1:nrow(clump)) {
  
  # Find all SNPs in LD (r2>0.6) with each clumped SNP
  
    system(paste0("module load plink/1.9b6 \
                   plink --bfile ", genotypeF, " --r2 --ld-window-kb 10000 --ld-window 2000 --ld-window-r2 0.6 --ld-snp ",
                  clump$SNP[i], " --out tmpld_", annot_name))
  
  # Read in the LD calculated from plink
  
  LD = read.table(paste0("tmpld_", annot_name, ".ld"), header = TRUE)
  
  # Turn into a genomic ranges object
  
  if (i==1) { 
    LDSNPs = GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP = LD$SNP_B, indexSNP = LD$SNP_A)
  } else {
    LDSNPs = c(GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP = LD$SNP_B, indexSNP = LD$SNP_A), LDSNPs)
  }
}

# Read in the eQTL data
eqtl = read.table(feqtl, header = TRUE)
eqtl.GR = GRanges(gsub("chr", "", eqtl$SNP_chr), IRanges(eqtl$SNP_start, eqtl$SNP_end))
mcols(eqtl.GR) = eqtl[, c(1:8,12:15)]

# Read in the BED file containing the annotation file
annot = read.table(bedfile, header = FALSE)
annotGR = GRanges(gsub("chr", "", annot$V1), IRanges(annot$V2, annot$V3), type = annot$V4)

# Find overlaps
olap = findOverlaps(annotGR, LDSNPs)
regionalSA_vars = unique(LDSNPs$indexSNP[subjectHits(olap)])
regionalAnnot = LDSNPs[which(!is.na(match(LDSNPs$indexSNP, regionalSA_vars)))]

# Outputting the number of loci that overlap with an HGE
cat('Number of loci that overlap with ', annot_name, ' is: ', length(unique(regionalAnnot$indexSNP)), '\n')

olap2 = findOverlaps(eqtl.GR, regionalAnnot)
regionalSA_eqtl_vars = unique(regionalAnnot$indexSNP[subjectHits(olap2)])
cat('Number of ', annot_name, ' overlapping loci that also have an eQTL: ', length(unique(regionalSA_eqtl_vars$indexSNP[subjectHits(olap2)])), '\n')
eqtlgenes = unique(eqtl.GR$gene_id[queryHits(olap2)])

# Remove the . annotation
eqtlgenes = sapply(eqtlgenes, function (x) {unlist(strsplit(x, ".", fixed = TRUE))[1]})

if (length(unique(regionalSA_eqtl_vars$indexSNP[subjectHits(olap2)])) > 0) {
  
  # Convert these genes to hgnc_id
  mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                 dataset = "hsapiens_gene_ensembl",
                 host = "feb2014.archive.ensembl.org")
  geneannot = getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                    filters = "ensembl_gene_id",
                    values = eqtlgenes,
                    mart = mart)
  cat('These eQTLs impact ', length(which(geneannot == "protein_coding")), ' protein-coding eGenes\n')
  write.csv(geneannot, file = paste0(outDir, "/rightHem_SA_", annot_name, "_lr_eqtl_genes.csv"), row.names = FALSE, quote = FALSE)
  
} else {
  print("There are not any CSA-associated variant - eQTL overlap")
}

if (length(regionalSA_vars) > 0) {

  m = matrix(NA, nrow = length(regionalSA_vars), ncol = 3)
  d = as.data.frame(m)
  colnames(d) = c("olap_snps", "eqtl_snps", "region")
  
  d$olap_snps = regionalSA_vars
  d$eqtl_snps[match(regionalSA_eqtl_vars, d$olap_snps)] = "Yes"
  d$region = "full_rightHem"
  d$olap_snps[match(regionalSA_eqtl_vars, d$olap_snps)]
  write.csv(d, file = paste0(outDir, "/rightHem_SA_", annot_name, "_lr_olap_snps.csv"),
            row.names = FALSE, quote = FALSE)
} else {
  print("There are not any CSA-associated variants - annotation overlap")
}

```

```{r regionalSA_left}

# All regional surface areas (LEFT HEM.)
# Start by getting all the clumped SNPs only for any regional surface area

surfareaind = grep("surface_le", phenoname) 
surfareaind = surfareaind[surfareaind != which(phenoname == "surface_le_Full")]

# Initiate a clump file, we will add other clump files to this one
clump = read.table(clumpfileloc[surfareaind[1]], header = TRUE)

first_region = strsplit(clumpfileloc[surfareaind[1]],"_",fixed=TRUE)[[1]][8]
clump$region = first_region

for (j in 2:length(surfareaind)) {
    
  if (file.exists(clumpfileloc[surfareaind[j]])) {
    tmp_clump = read.table(clumpfileloc[surfareaind[j]], header=TRUE)
    tmp_region_name = strsplit(clumpfileloc[surfareaind[j]],"_",fixed=TRUE)[[1]][8]
    tmp_clump$region = tmp_region_name
    clump = rbind(clump, tmp_clump, header=TRUE)
  }
}

clump = clump[clump$SNP!=TRUE,]

  # Loop over all SNPs
  for (i in 1:nrow(clump)) {
    
    # Find all SNPs in LD (r2>0.6) with each clumped SNP
    system(paste0("module load plink/1.9b6 \
                   plink --bfile ", genotypeF, " --r2 --ld-window-kb 10000 --ld-window 2000 --ld-window-r2 0.6 --ld-snp ",
                  clump$SNP[i], " --out tmpld_", annot_name))
    
    # Read in the LD calculated from plink
    LD = read.table(paste0("reg_tmpld_", annot_name, ".ld"), header = TRUE)
    
    # Turn into a genomic ranges object
    if (i==1) { 
      LDSNPs = GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP=LD$SNP_B, indexSNP=LD$SNP_A)
    } else {
      LDSNPs = c(GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP=LD$SNP_B, indexSNP=LD$SNP_A), LDSNPs)
    }
  }
  
# Make sure that all index SNPs are in this list
LDSNPs = c(GRanges(clump$CHR, IRanges(clump$BP, clump$BP), SNP = clump$SNP, indexSNP = clump$SNP), LDSNPs)
  
#save(LDSNPs,file="RegionalLDSNPs.Rdata")

# Find overlaps
olap = findOverlaps(annotGR, LDSNPs)
regionalSA_vars = unique(LDSNPs$indexSNP[subjectHits(olap)])
regionalAnnot = LDSNPs[which(!is.na(match(LDSNPs$indexSNP, regionalSA_vars)))]

# Outputting the number of loci that overlap with an HGE
cat('Number of loci that overlap with ', annot_name, ' is: ', length(unique(regionalAnnot$indexSNP)), '\n')

olap2 = findOverlaps(eqtl.GR, regionalAnnot)
regionalSA_eqtl_vars  =unique(regionalAnnot$indexSNP[subjectHits(olap2)])
cat('Number of ', annot_name, ' overlapping loci that also have an eQTL: ', length(unique(regionalAnnot$indexSNP[subjectHits(olap2)])), '\n')
eqtlgenes = unique(eqtl.GR$gene_id[queryHits(olap2)])
  
# Remove the . annotation
eqtlgenes = sapply(eqtlgenes, function (x) {unlist(strsplit(x, ".", fixed = TRUE))[1]})

if (length(unique(regionalAnnot$indexSNP[subjectHits(olap2)])) > 0) {
  
  # Convert these genes to hgnc_id
  mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                 dataset = "hsapiens_gene_ensembl",
                 host = "feb2014.archive.ensembl.org")
  geneannot = getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                    filters = "ensembl_gene_id", values = eqtlgenes,mart = mart)
  cat('These eQTLs impact ', length(which(geneannot == "protein_coding")), ' protein-coding eGenes\n')
  write.csv(geneannot, file = paste0(outDir, "/regionalSA_le_", annot_name, "_lr_eqtl_genes.csv"), row.names = FALSE, quote = FALSE)

} else {
  print("There are not any CSA-associated variant - eQTL overlap")
}

if (length(regionalSA_vars) > 0) {

  m = matrix(NA, nrow = length(regionalSA_vars), ncol = 3)
  d = as.data.frame(m)
  colnames(d) = c("olap_snps", "eqtl_snps", "region")
  
  d$olap_snps = regionalSA_vars
  d$eqtl_snps[match(regionalSA_eqtl_vars, d$olap_snps)] = "Yes"
  d$region = clump[match(regionalSA_vars, clump$SNP),]$region
  
  write.csv(d, file = paste0(outDir, "/regionalSA_le_", annot_name, "_lr_olap_snps.csv"),
            row.names = FALSE, quote = FALSE)
} else {
  print("There are not any CSA-associated variants - annotation overlap")
}

```

```{r regionalSA_right}

# All regional surface areas (RIGHT HEM.)
# Start by getting all the clumped SNPs only for any regional surface area

surfareaind = grep("surface_re", phenoname)
surfareaind = surfareaind[surfareaind != which(phenoname == "surface_re_Full")]

# Initiate a clump file, we will add other clump files to this one
clump = read.table(clumpfileloc[surfareaind[1]], header = TRUE)

first_region = strsplit(clumpfileloc[surfareaind[1]],"_",fixed=TRUE)[[1]][8]
clump$region = first_region

for (j in 2:length(surfareaind)) {
    
  if (file.exists(clumpfileloc[surfareaind[j]])) {
    tmp_clump = read.table(clumpfileloc[surfareaind[j]], header=TRUE)
    tmp_region_name = strsplit(clumpfileloc[surfareaind[j]],"_",fixed=TRUE)[[1]][8]
    tmp_clump$region = tmp_region_name
    clump = rbind(clump, tmp_clump, header=TRUE)
  }
}

clump = clump[clump$SNP!=TRUE,]

# Loop over all SNPs
for (i in 1:nrow(clump)) {
  
  # Find all SNPs in LD (r2>0.6) with each clumped SNP
    system(paste0("module load plink/1.9b6 \
                   plink --bfile ", genotypeF, " --r2 --ld-window-kb 10000 --ld-window 2000 --ld-window-r2 0.6 --ld-snp ",
                  clump$SNP[i], " --out tmpld_", annot_name))
  
  # Read in the LD calculated from plink
  LD = read.table(paste0("reg_tmpld_", annot_name, ".ld"), header = TRUE)
  
  # Turn into a genomic ranges object
  if (i==1) { 
    LDSNPs = GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP = LD$SNP_B, indexSNP = LD$SNP_A)
  } else {
    LDSNPs = c(GRanges(LD$CHR_B, IRanges(LD$BP_B, LD$BP_B), SNP = LD$SNP_B, indexSNP = LD$SNP_A), LDSNPs)
  }
}

# Make sure that all index SNPs are in this list
LDSNPs = c(GRanges(clump$CHR, IRanges(clump$BP, clump$BP), SNP = clump$SNP, indexSNP = clump$SNP), LDSNPs)

#save(LDSNPs,file="RegionalLDSNPs.Rdata")

# Find overlaps
olap = findOverlaps(annotGR, LDSNPs)
regionalSA_vars = unique(LDSNPs$indexSNP[subjectHits(olap)])
regionalAnnot = LDSNPs[which(!is.na(match(LDSNPs$indexSNP, regionalSA_vars)))]

# Outputting the number of loci that overlap with an HGE
cat('Number of loci that overlap with ', annot_name, ' is: ', length(unique(regionalAnnot$indexSNP)), '\n')

olap2 = findOverlaps(eqtl.GR, regionalAnnot)
regionalSA_eqtl_vars  =unique(regionalAnnot$indexSNP[subjectHits(olap2)])
cat('Number of ', annot_name, ' overlapping loci that also have an eQTL: ', length(unique(regionalAnnot$indexSNP[subjectHits(olap2)])), '\n')
eqtlgenes = unique(eqtl.GR$gene_id[queryHits(olap2)])

# Remove the . annotation
eqtlgenes = sapply(eqtlgenes, function (x) {unlist(strsplit(x, ".", fixed = TRUE))[1]})

if (length(unique(regionalAnnot$indexSNP[subjectHits(olap2)])) > 0) {
  
  # Convert these genes to hgnc_id
  mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                 dataset = "hsapiens_gene_ensembl",
                 host = "feb2014.archive.ensembl.org")
  geneannot = getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                    filters = "ensembl_gene_id", values = eqtlgenes, mart = mart)
  cat('These eQTLs impact ', length(which(geneannot == "protein_coding")), ' protein-coding eGenes\n')
  write.csv(geneannot, file = paste0(outDir, "/regionalSA_ri_", annot_name, "_lr_eqtl_genes.csv"), row.names = FALSE, quote = FALSE)
  
} else {
  print("There are not any CSA-associated variant - eQTL overlap")
}

if (length(regionalSA_vars) > 0) {

  m = matrix(NA, nrow = length(regionalSA_vars), ncol = 3)
  d = as.data.frame(m)
  colnames(d) = c("olap_snps", "eqtl_snps", "region")
  
  d$olap_snps = regionalSA_vars
  d$eqtl_snps[match(regionalSA_eqtl_vars, d$olap_snps)] = "Yes"
  d$region = clump[match(regionalSA_vars, clump$SNP),]$region
  
  write.csv(d, file = paste0(outDir, "/regionalSA_ri_", annot_name, "_lr_olap_snps.csv"),
            row.names = FALSE, quote = FALSE)
} else {
  print("There are not any CSA-associated variants - annotation overlap")
}

```

## Circular plot
                y = sample("Full left-hemisphere", 
                    "Pars opercularis",
                    "Lingual",
                    "Pars orbitalis",
                    "Cuneus", n, replace = T)
```{r}
set.seed(999)
n = 100
df = data.frame(sectors = sample(letters[1:8], n, replace = TRUE),
    x = rnorm(n), y = runif(n))

df = data.frame(sectors = sample(c("Neanderthal Depleted", 
                            "Foetal HGEs",
                            "Adult HGEs",
                            "Neanderthal Reintroduced",
                            "Human-specific Duplications",
                            "Full left-hemisphere", 
                            "Pars opercularis",
                            "Lingual",
                            "Pars orbitalis",
                            "Cuneus"), n, replace = T),
                x = rnorm(n), y = runif(n))

circos.par("track.height" = 0.1)
#circos.par(gap.after = c(rep(1, 23), 5, rep(1, 20), 5))
circos.initialize(df$sectors, x = df$x)

circos.track(df$sectors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, 
            CELL_META$cell.ylim[2] + mm_y(5), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})
col = rep(c("#FF0000", "#00FF00"), 5)
circos.trackPoints(df$sectors, df$x, df$y, col = col, pch = 16, cex = 0.5)
#circos.text(-1, 0.5, "text", sector.index = "Neanderthal Depleted", track.index = 1)
#bgcol = rep(c("#EFEFEF", "#CCCCCC"), 5)
#circos.trackHist(df$sectors, df$x, bin.size = 0.2, bg.col = bgcol, col = NA)

circos.link("Neanderthal Depleted", 0, "Pars opercularis", 0, h = 0.4, col = "green")
circos.link("c", c(-0.5, 0.5), "d", c(-0.5,0.5), col = "red",
    border = "blue", h = 0.2)
circos.link("e", 0, "g", c(-1,1), col = "green", border = "black", lwd = 2, lty = 2)
circos.clear()
```
```{r}
gwas_hit_snps=read.table("/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/eqtl/european_lr/results/results/gwas_hit_snps.txt")
colnames(gwas_hit_snps)=c("SNP", "A1", "A2", "FREQ1", "BETA",
                          "SE", "P", "N", "MARKER", "CHR", "BP", "INFO")
snp_region_list=read.table("/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/eqtl/european_lr/results/results/all_eqtl_regLeft.txt", header=T)

ultimate_super_df = merge(gwas_hit_snps, snp_region_list, by="SNP")

nondup_df = ultimate_super_df[!duplicated(ultimate_super_df), ]
ordered_df = nondup_df[order(nondup_df$P),]

ordered_df[ordered_df$SNP=="rs11423823",]

for_table = ordered_df[1:40,c("SNP", "A1", "FREQ1", 
                              "BETA", "SE", "P", "MARKER",
                              "brain_region", "annot", "eQTL")]

table_df = kable(for_table, row.names = F, format = "html", digits = 60, format.args = list(scientific = T), 
                 escape = F, align = "c",  table.attr = "style = \"color: black;\"") %>%
                  kable_styling(bootstrap_options = "striped", full_width = F, html_font = "Cambria", font_size = 15)
library(kableExtra)
library(knitr)
kable(table_df, format="latex", booktabs=TRUE) %>% 
  kable_styling(latex_options="scale_down")

kable(table_df, "latex", longtable = T, booktabs = T) %>%
  kable_styling(font_size = 7)

save_kable(table_df,"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/enigma_evol/enigma_evo/evolution/results/eqtl/european_lr/results/results/top20_snps_table.png")

circos_df = ultimate_super_df[,c("BETA", "P", "brain_region", "annot")]
circos_df1 = circos_df[c(3, 2, 1)]
colnames(circos_df1) = c("sectors", "x", "y")
circos_df2 = circos_df[c(4, 2, 1)]
colnames(circos_df2) = c("sectors", "x", "y")

df2 = rbind(circos_df1, circos_df2)

# make it smaller to test
df2 = df2[1:80,]

circos.par("track.height" = 0.1)
#circos.par$cell.padding[c(2,4)]=0
circos.initialize(df2$sectors, x = df2$x)
circos.track(df$sectors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, 
            CELL_META$cell.ylim[2] + mm_y(5), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})

circos.link("a", 0, "zoom_a", 0, h = 0.4, col = "green")
circos.link("c", c(-0.5, 0.5), "d", c(-0.5,0.5), col = "red",
    border = "blue", h = 0.2)
circos.link("e", 0, "b", c(-1,1), col = "green", border = "black", lwd = 2, lty = 2)

circos.link("a", get.cell.meta.data("cell.xlim", sector.index = "a"),
    "zoom_a", get.cell.meta.data("cell.xlim", sector.index = "zoom_a"),
    border = NA, col = "#00000020")
circos.link("b", c(zoom_df_b[1, 2], zoom_df_b[10, 2]),
    "zoom_b", get.cell.meta.data("cell.xlim", sector.index = "zoom_b"),
    rou1 = get.cell.meta.data("cell.top.radius", sector.index = "b"),
    border = NA, col = "#00000020")
circos.clear()
```


```{r}
sessionInfo()
#knitr::purl("overlap_analysis_eQTL.Rmd", "overlap_analysis_eQTL.R", documentation = 2) # save .Rmd as an .R file to submit to Grid
```
